File: snippets/cpp/01_search/binary_search.cpp
#include <vector>

int binary_search(const std::vector<int>& arr, int target) {
    int low = 0;
    int high = static_cast<int>(arr.size()) - 1;

    while (low <= high) {
        int mid = low + (high - low) / 2;
        if (arr[mid] == target) return mid;
        if (arr[mid] < target) {
            low = mid + 1;
        } else {
            high = mid - 1;
        }
    }

    return -1;
}

File: snippets/rust/01_search/binary_search.rs
pub fn binary_search(arr: &[i32], target: i32) -> Option<usize> {
    let mut low = 0usize;
    let mut high = arr.len();

    while low < high {
        let mid = low + (high - low) / 2;
        if arr[mid] == target {
            return Some(mid);
        }
        if arr[mid] < target {
            low = mid + 1;
        } else {
            high = mid;
        }
    }

    None
}
